### DISCUSSION TITLE
AIR NativeProcess - writing to standard input of child process (.NET) not working if the console app runs on .NET newer than 3.5

### DISCUSSION THREAD

**Discussion by daghendrik** [airsdk/Adobe-Runtime-Support#2796] [General]
- AIR version 33.1.633
- Platform: Windows

I am revisiting an old issue that I'm hoping there's a better solution for than creating a .NET 3.5 "middle man" console app to pipe messages between Adobe AIR and a .NET 4.8 app.

As described in this [Stack Overflow](https://stackoverflow.com/questions/14532442/air-nativeprocess-standardinput-not-connecting-to-app) post, Adobe AIR (via the NativeProcess API) has for some reason no problem writing to a console app running .NET 3.5, but if it tries to communicate with a console app running a newer version, for example .NET 4.8, the UTF bytes are never seen by the console app.

It's an old problem that I've been overcoming by writing a proxy/middle-man app in .NET 3.5 that forwards messages between AIR and my .NET 4.8 app, but I'm now hoping to be able to get rid of the .NET 3.5 dependency, which obsolete and no longer ships with Windows.

Is this something that is on the radar of the Adobe AIR developers, and if so, is it something that might be fixed in an upcoming version of AIR?

**Comment by marchbold** [comment]
Maybe a native extension might be a solution ? I have definitely been able to use more recent versions of .net in ANE's.

**Reply by ajwfrost** [reply]
ANEs are probably a good idea, but presumably would still need a wrapper (to translate from C-style calls into the managed .net code...)
Actually, another argument for the use of ANEs rather than NativeProcess launching a command-line app: I had run something like this recently, where the NativeProcess called an MSI file I think, and our IT folk sent me a message within minutes telling me of a policy violation and blocking my machine from our network! So, it may be worth avoiding doing things in a way that is sometimes also used by "bad actors" :-)

But in terms of the actual issue here, one of the team took a look and could reproduce this, the problem seems to be with the pipe that's set up to send data from AIR to the .NET process. The pipe appears to get set up and connect properly, but when we write to it, nothing happens, and when we added a call to "flush" then the thread just hung... so it may be something on the .NET side that's not correctly receiving the data from this.

We can look at this further - and if we hit a dead end, we can create a test app outside of AIR that mimics the behaviour, which would then at least enable other folk (e.g. on Stack Overflow) to take a look into it!

thanks

**Reply by daghendrik** [reply]
Thank you @ajwfrost, and sorry for the late reply.

I very much appreciate the feedback and the advice regarding the use of NativeProcess potentially being flagged as malicious.

Switching to ANEs is not currently an option for me, so if you and your team find this issue important enough, and have the resources, please look at it further and create that test app 👍
